agents/fts_stage/fts_agent.py

```diff
@@ -0,0 +1,304 @@
+            return True, self.pos
+        except:
+            return False, None
+
+    def move_to(self, position):
+        lims = self.LIMS
+        if position < lims[0] or position > lims[1]:
+            return False, 'Move out of bounds!'
+        M, B = self.TRANSLATE
+        stage_pos = position*M + B
+        cmd = ('MOVEABS X%.2f %s\n' % (stage_pos, self.speed_code))
+        self.send(cmd)
+        out = None
+        while out is None:
+            try:
+                # controller should block reply until move complete.
+                time.sleep(0.1)
+                out = self.read()
+            except TimeoutError:
+                continue         
+        return True, 'Move Complete'
+
+    def close(self):
+        self.comm.close()
+
+    
+
+class FTSMirrorAgent:
+    """
+    Agent for connecting to the FTS mirror control
+
+    Args:
+        ip_addr: IP address of Motion Controller
+        port: Port of Motion Controller
+        mode: 'acq': Start data acquisition on initialize
+        samp: default sampling frequency in Hz
+ 
+    """
+
+    def __init__(self, agent, ip_addr, port, mode=None, samp=2):
+
+        self.ip_addr = ip_addr
+        self.port = int(port)
+
+        self.stage = None
+        self.initialized = False
+        self.take_data = False
+
+        self.agent = agent
+        self.log = agent.log
+        self.lock = TimeoutLock()
+
+        if mode == 'acq':
+            self.auto_acq = True
+        else:
+            self.auto_acq = False
+        self.sampling_frequency = float(samp)
+
+        ### register the position feeds
+        agg_params = {
+            'frame_length' : 10*60, #[sec]
+        }
+
+        self.agent.register_feed('position',
+                                 record = True,
+                                 agg_params = agg_params,
+                                 buffer_time = 0)
+
+    def init_stage_task(self, session, params=None):
+        """init_stage_task(params=None)
+        Perform first time setup for communication with FTS stage.
+
+        Args:
+            params (dict): Parameters dictionary for passing parameters to
+                task.
+        """
+
+        if params is None:
+            params = {}
+        
+        if self.stage is not None and self.initialized:
+            return True, 'Stages already Initialized'
+           
+        self.log.debug("Trying to acquire lock")
+        with self.lock.acquire_timeout(timeout=0, job='init') as acquired:
+            # Locking mechanism stops code from proceeding if no lock acquired
+            if not acquired:
+                self.log.warn("Could not start init because {} is already" \
+                                "running".format(self.lock.job))
+                return False, "Could not acquire lock."
+            # Run the function you want to run
+            self.log.debug("Lock Acquired Connecting to Stages")
+            self.stage = FTSStage(self.ip_addr, self.port)
```

Comment:
I think this would run into the same scenario that you encountered in #164. A fix was implemented in #168 that you could adapt here to handle this case.

Reply:
Yup, that's it. The following return is still relevant too, and completes the task/process (by saying that it failed.)